Allow saving / loading checkpoints from cloud paths

agrinh opened the pull request:
Allows saving and loading checkpoints directly from cloud paths like
Amazon S3 (s3://) and Google Cloud Storage (gs://) by using fsspec.

Note: The user will have to install the relevant dependency for each
protocol. Otherwise fsspec will fail and specify which dependency is
missing.

agrinh commented:
I figured this would be a nice and pretty simple change, that will allow users (like me) to run coqui in environments where models aren't to be stored on the local disk. Simply adds the functions `save` and `load` (names can very well be discussed) which relies on fsspec to save and load from any path.

The new functions are a much simpler version of what's done in [pytorch-lightning.utilities.cloud_io](https://github.com/PyTorchLightning/pytorch-lightning/blob/b978d2a1f2f9cc9942971a6e252ccd1209a9269b/pytorch_lightning/utilities/cloud_io.py) and could be replaced if that dependency is ever adopted.

erogol commented on TTS/utils/io.py:
```diff
@@ -48,12 +49,23 @@ def copy_model_files(config, out_path, new_fields):
             )
 
 
+def load(path, **kwargs):
```
Maybe renaming them as `load_fsspec` and `save_fsspec` makes more sense ?

agrinh replied:
sure, that makes sense 🙂

erogol commented on requirements.txt:
```diff
@@ -24,3 +24,4 @@ mecab-python3==1.0.3
 unidic-lite==1.0.8
 # gruut+supported langs
 gruut[cs,de,es,fr,it,nl,pt,ru,sv]~=1.2.0
+fsspec
```
can you tag the working version too ?

agrinh replied:
definitely, but since these requirements are specified when installing as a dependency how about setting a lower bound?

erogol commented:
@agrinh  Would you also be interested in integrating Webdataset - https://github.com/webdataset/webdataset - for more cloud compat. ?

agrinh commented:
> @agrinh Would you also be interested in integrating Webdataset - https://github.com/webdataset/webdataset - for more cloud compat. ?

Looks pretty interesting, thanks for pointing it out but I'm not familiar with that lib. I'm currently on a setup more similar to tfds or e.g. the torchaudio datasets, but I've avoided tfds to not bring in tensorflow as a dependency, so this might be a nice alternative. I can take a look and get back with a PR / raise it in gitter later if it aligns with the workflow I have :)

agrinh commented:
I've now made it possible to have the remaining artifacts in cloud storage (at least I think I got everything) and added some documentation to the config dataclass. Some caveats:

- Logs are not written to the output path unless it is local. I'm not aware of a good existing way to log directly to a file-like object as opposed to a local path, and it didn't seem worth the effort right now since I'd expect users who run things in the cloud will have their logs collected elsewhere directly from stdout.
- I added a test to make sure urls are supported everywhere by prefixed the output path with "file://". This is more of a sanity check than anything, testing against some cloud storage in the future might be a good idea.

agrinh commented on TTS/bin/train_encoder.py:
```diff
@@ -212,12 +213,10 @@ def main(args):  # pylint: disable=redefined-outer-name
     try:
         main(args)
     except KeyboardInterrupt:
-        remove_experiment_folder(OUT_PATH)
```
The experiment folder is created once the first artifact is written, so this is no longer needed.

erogol replied:
how and where ? :)

erogol replied:
originally It does not wait for an artifact but a checkpoint file. 

So it helps to remove the folder if you ctrl+c before there is a checkpoint

agrinh replied:
`fsspec` actually creates the directory structure as needed when writing to paths, so it is created together with the first file. Though I'm not sure, I think that is when the config is written.

But you're right, it means that it no longer removes the folder if aborted after the config is written and before a checkpoint is created. Would you like me to try and re-implement the remove folder function but for fsspec? 

Personally I thought this behaviour really made sense, I this it's a good idea to keep the configs around even if aborted before a checkpoint is saved. E.g. even if I have no checkpoints there might be tensorboard logs with metrics I might want to examine.

erogol replied:
It makes the developer's life a lot harder and leaves too many marks after running the tests. I'd prefer to keep the behavior as before. In the end, you will have a checkpoint just after the first epoch.

agrinh replied:
Alright, I've restored it now.

erogol commented on TTS/utils/io.py:
```diff
@@ -37,23 +40,37 @@ def copy_model_files(config, out_path, new_fields):
     copy_config_path = os.path.join(out_path, "config.json")
     # add extra information fields
     config.update(new_fields, allow_new=True)
-    config.save_json(copy_config_path)
```
if you don't get lazy you can add save_fsspec to https://github.com/coqui-ai/coqpit :) 
Otherwise, it is better to use 

```python
config.to_json()
# or
config.to_dict()
```

agrinh replied:
Oh but I am lazy :D

Jokes aside, I will for sure make a PR to allow Coqpit to save with fsspec once this is in, but I figured adding this small bit here for now has the advantage of not depending on a dependency update so it's easier to test etc. For now I switched to `config.to_dict()`

erogol commented on TTS/utils/io.py:
```diff
@@ -37,23 +40,37 @@ def copy_model_files(config, out_path, new_fields):
+                    shutil.copyfileobj(source_file, target_file)
+
+
+def load_fsspec(path, **kwargs):
```
Can you also use type annotations for the new functions?

agrinh replied:
Will do! Unfortunately type annotations for `**kwargs` and the torch state object won't be very informative, but at least path will be.

erogol commented on tests/tts_tests/test_tacotron2_train_fsspec_path.py:
```diff
@@ -0,0 +1,55 @@
+import glob
```
How is it different than a regular training test except for the `file` paths ?

agrinh replied:
It's exactly the same :) I could also add "file://" paths to the original one (without the `fsspec_path` suffix), do you have a preference? This will have a test execution time penalty but it might perhaps be easier to tell if `fsspec` is the culprit.

erogol commented on TTS/utils/generic_utils.py:
```diff
@@ -58,28 +58,15 @@ def get_commit_hash():
     return commit
 
 
-def create_experiment_folder(root_path, model_name):
```
I guess this is what you mean above but not sure if it solves the case I explained above

agrinh replied:
Yeah! I replied above 🙂

agrinh commented:
@erogol Thanks for the review! I've addressed the questions I fixed right away and will wait for input regarding the others

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/coqui-ai/TTS?pullRequest=683) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/coqui-ai/TTS?pullRequest=683) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/coqui-ai/TTS?pullRequest=683) it.</sub>

agrinh commented:
@erogol I have no way of inspecting why the failure in your snyk test occurs. Please let me know if there's anything else I need to fix.
